test/unit/BPool/BPool_Unbind.t.sol

```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.25;
+
+import {BPoolBase} from './BPoolBase.sol';
+
+import {IERC20} from '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+import {IBPool} from 'interfaces/IBPool.sol';
+
+contract BPool_Unbind is BPoolBase {
+  address public secondToken = makeAddr('secondToken');
+
+  function setUp() public virtual override {
+    super.setUp();
+    vm.mockCall(secondToken, abi.encodePacked(IERC20.transferFrom.selector), abi.encode());
+  }
+
+  function test_WhenReentrancyLockIsNOTSetRevertWhen_CallerIsNOTController(address _caller) external {
+    // it should revert
+    vm.assume(_caller != deployer);
+    vm.prank(_caller);
+    vm.expectRevert(IBPool.BPool_CallerIsNotController.selector);
+    bPool.unbind(token);
+  }
+
+  modifier whenCallerIsController() {
+    vm.startPrank(deployer);
+    _;
+  }
+
+  function test_WhenReentrancyLockIsNOTSetRevertWhen_TokenIsNotBound() external whenCallerIsController {
+    vm.expectRevert(IBPool.BPool_TokenNotBound.selector);
+    // it should revert
+    bPool.unbind(token);
+  }
+
+  modifier whenTokenIsBound() {
+    _setRecord(token, IBPool.Record({bound: true, index: 0, denorm: tokenWeight}));
+    bPool.set__totalWeight(tokenWeight);
+    address[] memory tokens = new address[](1);
+    tokens[0] = token;
+    bPool.set__tokens(tokens);
+    _;
+  }
+
+  function test_WhenReentrancyLockIsNOTSetRevertWhen_PoolIsFinalized() external whenCallerIsController whenTokenIsBound {
+    bPool.set__finalized(true);
+    // it should revert
+    vm.expectRevert(IBPool.BPool_PoolIsFinalized.selector);
+    bPool.unbind(token);
+  }
+
+  function test_WhenReentrancyLockIsNOTSetWhenTokenIsLastOnTheTokensArray()
+    external
+    whenCallerIsController
+    whenTokenIsBound
+  {
+    // it sets reentrancy lock
+    bPool.expectCall__setLock(_MUTEX_TAKEN);
+    // it calls _pushUnderlying
+    bPool.expectCall__pushUnderlying(token, deployer, tokenBindBalance);
+
+    // it emits LOG_CALL event
+    vm.expectEmit();
+    bytes memory _data = abi.encodeWithSelector(IBPool.unbind.selector, token);
+    emit IBPool.LOG_CALL(IBPool.unbind.selector, deployer, _data);
+    bPool.unbind(token);
+
+    // it removes the token record
+    assertFalse(bPool.call__records(token).bound);
+    // it pops from the array
+    assertEq(bPool.getNumTokens(), 0);
+    // it decreases the total weight
+    assertEq(bPool.call__totalWeight(), 0);
+  }
+
+  function test_WhenReentrancyLockIsNOTSetWhenTokenIsNOTLastOneInTheArray()
+    external
+    whenCallerIsController
+    whenTokenIsBound
+  {
+    bPool.bind(secondToken, tokenBindBalance, tokenWeight);
```

Comment:
this is kinda forbidden but well, why cannot we use some mocked method to add another token to the array?